File: solutions/443. String Compression.cpp
class Solution {
 public:
  int compress(std::vector<char>& chars) {
    AddSentinel_(chars);
    int compressed_end = 0;
    int count = 0;
    int prev_c = SENTINEL_;
    for (const char c : chars) {
      if (c != prev_c && prev_c != SENTINEL_) {
        chars.at(compressed_end++) = prev_c;
        if (CanBeCompressed_(count)) {
          /* from MSB */
          for (char d : std::to_string(count)) {
            chars.at(compressed_end++) = d;
          }
        }
        count = 0;
      }
      count++;
      prev_c = c;
    }

    return compressed_end;
  }

 private:
  const char SENTINEL_ = '\0';

  // To eliminate the case which the last kind of characters
  // are not compressed.
  void AddSentinel_(std::vector<char>& chars) const {
    chars.push_back(SENTINEL_);
  }

  bool CanBeCompressed_(const int count) const {
    return count > 1;
  }
};

File: solutions/443. String Compression.rs
impl Solution {
    pub fn compress(chars: &mut Vec<char>) -> i32 {
        Self::add_sentinel(chars);
        let mut compressed_end: usize = 0;
        let mut count: u32 = 0;
        let mut prev_c: char = '\0';
        for i in 0..chars.len() {
            let c: char = chars[i];
            if c != prev_c && prev_c != '\0' {
                chars[compressed_end] = prev_c;
                compressed_end += 1;
                if Self::can_be_compressed(count) {
                    for d in count.to_string().chars() {
                        chars[compressed_end] = d;
                        compressed_end += 1;
                    }
                }
                count = 0;
            }
            count += 1;
            prev_c = c;
        }
        compressed_end as i32
    }

    fn add_sentinel(chars: &mut Vec<char>) {
        chars.push('\0')
    }

    fn can_be_compressed(count: u32) -> bool {
        count > 1
    }
}
